src/shared/tpm2-util.c

```diff
@@ -458,24 +458,161 @@ static int tpm2_make_primary(
+        assert(a->hash == b->hash);
 
-        *ret = (TPML_PCR_SELECTION) {
+        uint32_t maska = tpm2_tpms_pcr_selection_to_mask(a);
+        uint32_t maskb = tpm2_tpms_pcr_selection_to_mask(b);
+        *a = tpm2_tpms_pcr_selection_from_mask(maska & ~maskb, a->hash);
+}
+
+static unsigned tpm2_tpms_pcr_selection_weight(const TPMS_PCR_SELECTION *s) {
+        return (unsigned) __builtin_popcount(tpm2_tpms_pcr_selection_to_mask(s));
+}
+
+#define tpm2_tpms_pcr_selection_empty(s) (tpm2_tpms_pcr_selection_weight(s) == 0)
+
+#define FOREACH_PCR_IN_TPMS_PCR_SELECTION(pcr, tpms)                    \
+        FOREACH_PCR_IN_MASK(pcr, tpm2_tpms_pcr_selection_to_mask(tpms))
+
+#define FOREACH_TPMS_PCR_SELECTION_IN_TPML_PCR_SELECTION(tpms, tpml)    \
+        for (TPMS_PCR_SELECTION *tpms = (tpml)->pcrSelections;          \
+             (uint32_t)(tpms - (tpml)->pcrSelections) < (tpml)->count;  \
+             tpms++)
+
+#define FOREACH_PCR_IN_TPML_PCR_SELECTION(pcr, tpms, tpml)              \
+        FOREACH_TPMS_PCR_SELECTION_IN_TPML_PCR_SELECTION(tpms, tpml)    \
+                FOREACH_PCR_IN_TPMS_PCR_SELECTION(pcr, tpms)            \
+
+static char *tpm2_tpms_pcr_selection_to_string(const TPMS_PCR_SELECTION *s) {
+        const char *algstr = strna(tpm2_hash_alg_to_string(s->hash));
+
+        _cleanup_free_ char *maskstr = tpm2_pcr_mask_to_string(tpm2_tpms_pcr_selection_to_mask(s));
+        if (!maskstr)
+                return NULL;
+
+        return strjoin(algstr, "(", maskstr, ")");
+}
+
+/* Utility functions for TPML_PCR_SELECTION. */
+
+static TPMS_PCR_SELECTION *tpm2_tpml_pcr_selection_get_tpms_pcr_selection(TPML_PCR_SELECTION *l, TPMI_ALG_HASH hash) {
+        FOREACH_TPMS_PCR_SELECTION_IN_TPML_PCR_SELECTION(s, l)
+                if (s->hash == hash)
+                        return s;
+        return NULL;
+}
+
+uint32_t tpm2_tpml_pcr_selection_to_mask(const TPML_PCR_SELECTION *l, TPMI_ALG_HASH hash) {
+        TPMS_PCR_SELECTION *s = tpm2_tpml_pcr_selection_get_tpms_pcr_selection((TPML_PCR_SELECTION*) l, hash);
+        return s ? tpm2_tpms_pcr_selection_to_mask(s) : 0;
+}
+
+TPML_PCR_SELECTION tpm2_tpml_pcr_selection_from_mask(uint32_t mask, TPMI_ALG_HASH hash) {
+        return (TPML_PCR_SELECTION) {
                 .count = 1,
-                .pcrSelections[0] = {
-                        .hash = bank,
-                        .sizeofSelect = 3,
-                        .pcrSelect[0] = mask & 0xFF,
-                        .pcrSelect[1] = (mask >> 8) & 0xFF,
-                        .pcrSelect[2] = (mask >> 16) & 0xFF,
-                }
+                .pcrSelections[0] = tpm2_tpms_pcr_selection_from_mask(mask, hash),
         };
 }
 
+static unsigned tpm2_tpml_pcr_selection_weight(const TPML_PCR_SELECTION *l) {
+        unsigned weight = 0;
+        for (unsigned i = 0; i < l->count; i++)
+                weight += tpm2_tpms_pcr_selection_weight(&l->pcrSelections[i]);
```

Comment:
this can accumulate 24 at max, right?

maybe add an assert() that tells readers and compiler that this cannot possibly overflow

i.e. :+1: 
```
unsigned w = tpm2_tpms_pcr_selection_weight(&l->pcrSelections[i]);
assert(weight <= UINT_MAX - w);
```

Reply:
ack